runtime/libia2/destructors.c

```diff
@@ -0,0 +1,104 @@
+#include <elf.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <ia2_internal.h>
+
+/* UBSan doesn't like this access, so opt it out of sanitization. */
+__attribute__((no_sanitize("undefined")))
+static uint32_t ubsan_access_phdr_type(Elf64_Phdr *phdr, int i) {
+    return phdr[i].p_type;
+}
+
+// TODO: visibility may not be necessary anymore since this is only called from libia2.a
+__attribute__((visibility("default"))) void ia2_setup_destructors(const Elf64_Ehdr *ehdr, int pkey, void *dtor_callgate, void *dtor_ptr, struct FinalizerInfo *finalizers) {
+  int res = 0;
+  Elf64_Phdr *phdr = (Elf64_Phdr *)((uint8_t *)ehdr + ehdr->e_phoff);
+  assert(sizeof(Elf64_Phdr) == ehdr->e_phentsize);
+  Elf64_Phdr *dynamic_phdr = NULL;
+  for (int i = 0; i < ehdr->e_phnum; i++) {
+    if (ubsan_access_phdr_type(phdr, i) == PT_DYNAMIC) {
+      dynamic_phdr = &phdr[i];
+      break;
+    }
+  }
+  if (!dynamic_phdr) {
+    return;
+  }
+  Elf64_Dyn *dynamic =
+      (Elf64_Dyn *)((uintptr_t)ehdr + dynamic_phdr->p_vaddr);
+  void *rewrite_base = NULL;
+  void *finalizers_region =
+      (void *)IA2_ROUND_DOWN((uintptr_t)finalizers, PAGE_SIZE);
+  size_t finalizers_region_size =
+      IA2_ROUND_DOWN(((uintptr_t)finalizers) + sizeof(finalizers) + PAGE_SIZE,
```

Comment:
Is this a round-up? In that case, should it be `(uintptr_t)finalizers) + sizeof(finalizers) + PAGE_SIZE - 1`?

Reply:
same as above